docs/USAGE-GUIDE.md

```diff
@@ -1929,6 +1929,95 @@ s2n-tls will consider all allocated memory available for use, and will attempt t
+```
+uint8_t early_data[] = "early data to send";
+ssize_t total_data_sent = 0, len = sizeof(early_data);
+while (total_data_sent < len) {
+    ssize_t data_sent = 0;
+    int r = s2n_send_early_data(client_conn, early_data + total_data_sent,
+            len - total_data_sent, &data_sent, &blocked);
+    total_data_sent += data_sent;
+    if (r == S2N_SUCCESS) {
+        break;
+    } else if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+        exit(1);
+    }
+}
+while (s2n_negotiate(client_conn, &blocked) != S2N_SUCCESS) {
+    if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+        exit(1);
+    }
+}
+```
+
+### Receiving early data
+
+To receive early data, your application should call **s2n_recv_early_data** before it calls **s2n_negotiate**. Like other S2N IO functions, **s2n_recv_early_data** can potentially fail repeatedly with a blocking error before it eventually succeeds: see [I/O Functions](#io-functions) for more information.
+
+Once **s2n_recv_early_data** has been called once, it must be called until it returns success. If an application stops calling **s2n_recv_early_data** early, some early data may be left unread and cause later calls to **s2n_negotiate** to fail. Calling **s2n_recv_early_data** again after it returns success is possible but does nothing.
+
+Once a server has read all early data, you call **s2n_negotiate** to complete the handshake just as you would for a handshake that did not include early data.
+
+For example:
+```
+uint8_t early_data[MAX_EARLY_DATA] = { 0 };
+ssize_t total_data_recv = 0, data_recv = 0;
+while (s2n_recv_early_data(conn, early_data + total_data_recv, MAX_EARLY_DATA - total_data_recv,
+        &data_recv, &blocked) != S2N_SUCCESS) {
+    total_data_recv += data_recv;
+    if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+        exit(1);
+    }
+}
+while (s2n_negotiate(conn, &blocked) != S2N_SUCCESS) {
+    if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+        exit(1);
+    }
+}
+```
+
+### Adding anti-replay protection
+s2n-tls does not include anti-replay protection automatically. Effective anti-replay protection for a multi-server application requires an external state shared by all servers. Without shared state, an attacker can capture early data originally sent to server A and successfully replay it against server B.
+
+Instead, s2n-tls provides a callback to hook into your anti-replay solution. The **s2n_early_data_cb** can be used to accept or reject client early data requests. The callback can be configured by using **s2n_config_set_early_data_cb**.
+
+Using the **s2n_offered_early_data** pointer offered by the callback, call **s2n_offered_early_data_reject** and **s2n_offered_early_data_accept** to accept or reject the client request to use early data. The callback can be implemented asynchronously if desired: returning without calling either method will cause the connection to block until the early data is either rejected or accepted.
```

Comment:
wonder if an example of the usage will be valuable here.

`returning without calling either method will cause the connection to block` does that mean one of the functions should be called? what happens if the callback is async, does that mean it blocks?